src/pyhf/infer/test_statistics.py

```diff
@@ -56,7 +56,10 @@ def qmu(mu, data, pdf, init_pars, par_bounds):
         data, pdf, init_pars, par_bounds, return_fitted_val=True
     )
     qmu = fixed_poi_fit_lhood_val - unconstrained_fit_lhood_val
-    qmu = tensorlib.where(
-        muhatbhat[pdf.config.poi_index] > mu, tensorlib.astensor(0.0), qmu
+    # Required for backend uniformity (JAX differs from others)
+    qmu = tensorlib.atleast_1d(
```

Comment:
can you explain this in more detail, why is jax different?

Reply:
@lukasheinrich 

```python
import pyhf
from pyhf.infer.mle import fixed_poi_fit, fit
from pyhf.exceptions import UnspecifiedPOI


def _qmu(mu, data, pdf, init_pars, par_bounds):
    if pdf.config.poi_index is None:
        raise UnspecifiedPOI(
            'No POI is defined. A POI is required for profile likelihood based test statistics.'
        )

    tensorlib, optimizer = pyhf.get_backend()
    mubhathat, fixed_poi_fit_lhood_val = fixed_poi_fit(
        mu, data, pdf, init_pars, par_bounds, return_fitted_val=True
    )
    muhatbhat, unconstrained_fit_lhood_val = fit(
        data, pdf, init_pars, par_bounds, return_fitted_val=True
    )
    qmu = fixed_poi_fit_lhood_val - unconstrained_fit_lhood_val
    qmu = tensorlib.where(
        muhatbhat[pdf.config.poi_index] > mu, tensorlib.astensor(0.0), qmu
    )[0]
    return tensorlib.clip(qmu, 0, max_value=None)


def main():
    for backend in ["numpy", "pytorch", "tensorflow", "jax"]:
        print(f"\nUsing {backend} backend")
        pyhf.set_backend(backend)
        model = pyhf.simplemodels.hepdata_like(
            signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
        )
        observations = [51, 48]
        data = pyhf.tensorlib.astensor(observations + model.config.auxdata)
        test_mu = 1.0
        init_pars = model.config.suggested_init()
        par_bounds = model.config.suggested_bounds()
        # This (enforcing a non-0-d shape) and then taking the 0th element works
        # q_mu = pyhf.infer.test_statistics.qmu(
        #     test_mu, data, model, init_pars, par_bounds
        # )

        # This doesn't
        q_mu = _qmu(test_mu, data, model, init_pars, par_bounds)
        print(f"q_mu for backend {backend} is: {q_mu}")


if __name__ == '__main__':
    main()
```

gives

```
$ python example_qmu_error.py 

Using numpy backend
q_mu for backend numpy is: 3.938244920380498

Using pytorch backend
q_mu for backend pytorch is: 3.938140869140625

Using tensorflow backend
q_mu for backend tensorflow is: 3.938232421875

Using jax backend
/home/feickert/.venvs/pyhf-CPU/lib/python3.7/site-packages/jax/lib/xla_bridge.py:123: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Traceback (most recent call last):
  File "example_qmu_error.py", line 46, in <module>
    main()
  File "example_qmu_error.py", line 41, in main
    q_mu = _qmu(test_mu, data, model, init_pars, par_bounds)
  File "example_qmu_error.py", line 22, in _qmu
    )[0]
  File "/home/feickert/.venvs/pyhf-CPU/lib/python3.7/site-packages/jax/numpy/lax_numpy.py", line 2786, in _rewriting_take
    return _gather(arr, treedef, static_idx, dynamic_idx)
  File "/home/feickert/.venvs/pyhf-CPU/lib/python3.7/site-packages/jax/numpy/lax_numpy.py", line 2793, in _gather
    indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
  File "/home/feickert/.venvs/pyhf-CPU/lib/python3.7/site-packages/jax/numpy/lax_numpy.py", line 2878, in _index_to_gather
    idx = _canonicalize_tuple_index(len(x_shape), idx)
  File "/home/feickert/.venvs/pyhf-CPU/lib/python3.7/site-packages/jax/numpy/lax_numpy.py", line 3129, in _canonicalize_tuple_index
    raise IndexError(msg.format(len_without_none, arr_ndim))
IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0.
```

Reply:
hm do we understand why, though?

Reply:
on master I get this

```
Using numpy backend
qmu is [3.93824492] and of type <class 'numpy.ndarray'> with shape (1,)
q_mu for backend numpy is: 3.938244920380498

Using pytorch backend
qmu is tensor([3.9380]) and of type <class 'torch.Tensor'> with shape (1,)
q_mu for backend pytorch is: 3.93798828125

Using tensorflow backend
2020-07-21 23:50:26.714860: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-07-21 23:50:26.731688: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fa2c3aafc50 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-21 23:50:26.731708: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:tensorflow:From /Users/lukasheinrich/Code/pyhfdev/dev/pyhfdevenv/lib/python3.7/site-packages/tensorflow/python/ops/math_grad.py:297: setdiff1d (from tensorflow.python.ops.array_ops) is deprecated and will be removed after 2018-11-30.
Instructions for updating:
This op will be removed after the deprecation date. Please switch to tf.sets.difference().
WARNING:tensorflow:From /Users/lukasheinrich/Code/pyhfdev/dev/pyhfdevenv/lib/python3.7/site-packages/tensorflow/python/ops/array_grad.py:644: _EagerTensorBase.cpu (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.identity instead.
qmu is [3.9382324] and of type <class 'tensorflow.python.framework.ops.EagerTensor'> with shape (1,)
q_mu for backend tensorflow is: 3.938232421875

Using jax backend
/Users/lukasheinrich/Code/pyhfdev/dev/pyhfdevenv/lib/python3.7/site-packages/jax/lib/xla_bridge.py:125: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
qmu is [3.93824492] and of type <class 'jax.interpreters.xla.DeviceArray'> with shape (1,)
q_mu for backend jax is: 3.938244920363122
```

Reply:
it looks like it's some version of #964

Reply:
```
import pyhf
from pyhf.infer.mle import fixed_poi_fit


def fitshape(mu, data, pdf, init_pars, par_bounds):
    _, fixed_poi_fit_lhood_val = fixed_poi_fit(
        mu, data, pdf, init_pars, par_bounds, return_fitted_val=True
    )
    print(fixed_poi_fit_lhood_val)


def main():
    for backend in ["numpy", "pytorch", "tensorflow", "jax"]:
        print(f"\nUsing {backend} backend")
        pyhf.set_backend(backend)
        model = pyhf.simplemodels.hepdata_like(
            signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
        )
        observations = [51, 48]
        data = observations + model.config.auxdata
        fitshape(1.0, data, model, model.config.suggested_init(), model.config.suggested_bounds())


if __name__ == '__main__':
    main()
```

somehow for jax the return value of the fit is a scalar and not a 1-d tensor

Reply:
> on master I get this

Yeah, but on `master` _everything_ is a tensor as we enforce minimum shape of `(1,)`.